### PR TITLE
[SPARK-36726] Upgrade Parquet to 1.12.1

### PR DESCRIPTION
**Pull request by sunchao**
### What changes were proposed in this pull request?

Upgrade Apache Parquet to 1.12.1

### Why are the changes needed?

Parquet 1.12.1 contains the following bug fixes:
- PARQUET-2064: Make Range public accessible in RowRanges
- PARQUET-2022: ZstdDecompressorStream should close `zstdInputStream`
- PARQUET-2052: Integer overflow when writing huge binary using dictionary encoding
- PARQUET-1633: Fix integer overflow
- PARQUET-2054: fix TCP leaking when calling ParquetFileWriter.appendFile
- PARQUET-2072: Do Not Determine Both Min/Max for Binary Stats 
- PARQUET-2073: Fix estimate remaining row count in ColumnWriteStoreBase
- PARQUET-2078: Failed to read parquet file after writing with the same

In particular PARQUET-2078 is a blocker for the upcoming Apache Spark 3.2.0 release.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests + a new test for the issue in SPARK-36696

### CODE REVIEW

**Comment by srowen** [comment]
This blocks 3.2.0 right? 
Maybe it's not 100% published just yet?
`[error] sbt.librarymanagement.ResolveException: Error downloading org.apache.parquet:parquet-column:1.12.1`

**Comment by sunchao** [comment]
Oops I missed a few places where the Apache staging repo needs to be added. Trying it again.

**Comment by sunchao** [comment]
> blocks 3.2.0 right?

Yes this blocks 3.2.0 - I'm testing RC0 here

**Comment by dongjoon-hyun** [comment]
Thank you, @sunchao !

**Review by viirya — COMMENTED** [review]
Thanks for this work! We also need to update dependency manifest.

**Comment by sunchao** [comment]
Thanks, will do in the next run. I verified this PR with the problematic file in [SPARK-36696](https://issues.apache.org/jira/browse/SPARK-36696), and it now works:

```scala
  test("SPARK-36726: test incorrect Parquet row group file offset") {
    readParquetFile("/tmp/example.parquet") { df =>
      assert(df.count() == 3650)
    }
  }
```

File: pom.xml
Comment (srowen) [inline pom.xml:312]: This is OK to test - we'd want to wait until it's officially released and then remove it before merging

**Comment by dongjoon-hyun** [comment]
cc @gengliangwang

File: pom.xml
Comment (gengliangwang) [reply pom.xml:312]: +1

**Comment by gengliangwang** [comment]
@sunchao  Thank you for making this PR right after the Parquet 1.12.1 RC0!

**Review by dongjoon-hyun — COMMENTED** [review]
Please regenerate the dependency manifests.
```
dev/test-dependencies.sh --replace-manifest
```

**Comment by holdenk** [comment]
I know this is covered by UT in Parquet, but would it make sense to cover the behaviour which was a regression inside of Spark UT as well?

**Comment by sunchao** [comment]
@holdenk yes this is a good point, just added. Also updated the dependency manifest files.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala
Comment (dbtsai) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:860]: should we test for both vec/non vec code path?

Comment (sunchao) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:860]: I think `readParquetFile` already takes care of that, see `FileBasedDataSourceTest.readFile`.

Comment (dbtsai) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:860]: I thought vectorized is enabled by default, and non-vectorized path will not be tested.  `  protected def readFile(path: String, testVectorized: Boolean = true)`, so we need

```
  Seq(true, false).foreach { vec =>
      readParquetFile(testFile("file.parquet"), vec) { df =>
         ...
     }
  }
```

**Comment by dongjoon-hyun** [comment]
BTW, I assumed that the last tests passed with RC1 instead of RC0.
@sunchao . Is there a way to check that?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala
Comment (sunchao) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:860]: This is how `FileBasedDataSourceTest.readFile` is defined:

```scala
  protected def readFile(path: String, testVectorized: Boolean = true)
      (f: DataFrame => Unit): Unit = {
    withSQLConf(vectorizedReaderEnabledKey -> "false") {
      f(spark.read.format(dataSourceName).load(path.toString))
    }
    if (testVectorized) {
      withSQLConf(vectorizedReaderEnabledKey -> "true") {
        f(spark.read.format(dataSourceName).load(path.toString))
      }
    }
  }
```
so to me it tests both paths. Did I miss anything?

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:860]: I am surprised that the non-vectorized reader is always tested with/without testVectorized enabled...
```
  protected def readFile(path: String, testVectorized: Boolean = true)
      (f: DataFrame => Unit): Unit = {
    withSQLConf(vectorizedReaderEnabledKey -> "false") {
      f(spark.read.format(dataSourceName).load(path.toString))
    }
    if (testVectorized) {
      withSQLConf(vectorizedReaderEnabledKey -> "true") {
        f(spark.read.format(dataSourceName).load(path.toString))
      }
    }
  }
```

**Comment by sunchao** [comment]
@dongjoon-hyun yes I manually checked https://repository.apache.org/content/repositories/staging/org/apache/parquet/ and verified it has the latests bits before kicking off the CI here. I also synced with @shangxinli who's the release manager on this.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala
Comment (sunchao) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:860]: Yes, I guess it makes sense since in most cases we'd want to test both code paths and this is convenient for that.

Comment (dbtsai) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:860]: You are right. I'm surprised that non-vectorized path is always tested. In many tests, we specifically test both paths, and they are not necessary given the non-vectorized will always be tested.

**Review by dbtsai — APPROVED** [review]
LGTM. Pending CI.

**Review by holdenk — APPROVED** [review]
+1 (pending ci)

**Comment by dbtsai** [comment]
Merged into master/branch-3.2. Thanks. 

cc @gengliangwang

**Comment by gengliangwang** [comment]
Late LGTM. Thank you, @sunchao !

**Pull request closed without merging**
